src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentFactoryBase.cs

```diff
@@ -0,0 +1,139 @@
+
+            if (textBufferFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textBufferFactory));
+            }
+
+            if (textDocumentFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textDocumentFactory));
+            }
+
+            if (filePathProvider is null)
+            {
+                throw new ArgumentNullException(nameof(filePathProvider));
+            }
+
+            _contentTypeRegistry = contentTypeRegistry;
+            _textBufferFactory = textBufferFactory;
+            _textDocumentFactory = textDocumentFactory;
+            _fileUriProvider = filePathProvider;
+        }
+
+        private IContentType LanguageLSPContentType
+        {
+            get
+            {
+                if (_languageLSPContentType == null)
+                {
+                    var contentType = _contentTypeRegistry.GetContentType(LanguageContentTypeName);
+                    _languageLSPContentType = ConvertToLSPContentType(contentType);
+                }
+
+                return _languageLSPContentType;
+            }
+        }
+
+        /// <summary>
+        /// Converts registered document content type to one that extends "code-languageserver-base" and
+        /// CodeRemoteContentDefinition.CodeRemoteContentTypeName if needed.
+        /// </summary>
+        /// <param name="contentType">Content type registered with IContentTypeRegistryService</param>
+        /// <returns>Corresponding LSP/remote content type</returns>
+        protected virtual IContentType ConvertToLSPContentType(IContentType contentType) => contentType;
+
+        public override bool TryCreateFor(ITextBuffer hostDocumentBuffer, out VirtualDocument virtualDocument)
+        {
+            if (hostDocumentBuffer is null)
+            {
+                throw new ArgumentNullException(nameof(hostDocumentBuffer));
+            }
+
+            if (!hostDocumentBuffer.ContentType.IsOfType(HostDocumentContentTypeName))
+            {
+                // Another content type we don't care about.
+                virtualDocument = null;
+                return false;
+            }
+
+            var hostDocumentUri = _fileUriProvider.GetOrCreate(hostDocumentBuffer);
+
+            // E.g. Index.cshtml => Index.cshtml__virtual.html (for html), similar for other languages
+            var virtualLanguageFilePath = hostDocumentUri.GetAbsoluteOrUNCPath() + LanguageFileNameSuffix;
+            var virtualLanguageUri = new Uri(virtualLanguageFilePath);
+
+            var languageBuffer = _textBufferFactory.CreateTextBuffer();
+            _fileUriProvider.AddOrUpdate(languageBuffer, virtualLanguageUri);
+            languageBuffer.Properties.AddProperty(LSPConstants.ContainedLanguageMarker, true);
+
+            if (!(AdditionalLanguageBufferProperties is null))
+            {
+                foreach (KeyValuePair<object, object> keyValuePair in AdditionalLanguageBufferProperties)
```

Comment:
```suggestion
                foreach (var keyValuePair in AdditionalLanguageBufferProperties)
```

Reply:
Oh year, I forget to do vars :) Changed.

---
In reply to: [451970522](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r451970522) [](ancestors = 451970522)